content/workspace/developers/ai-features/mcp-tools.md

```diff
@@ -0,0 +1,235 @@
+                                system_content += f" - {param_desc}"
+                            system_content += "\n"
+    
+    # Create OpenAI function definition for MCP tools
+    functions = []
+    if request.tools:
+        functions.append({
+            "name": "execute_agent_tool",
+            "description": "Execute an MCP tool to retrieve data",
+            "parameters": {
+                "type": "object",
+                "properties": {
+                    "server_id": {
+                        "type": "string",
+                        "description": "The ID of the MCP server",
+                        "enum": list(set(
+                            getattr(tool, "server_id", "unknown")
+                            for tool in request.tools
+                        )),
+                    },
+                    "tool_name": {
+                        "type": "string",
+                        "description": "The name of the tool to execute",
+                        "enum": [tool.name for tool in request.tools],
+                    },
+                    "parameters": {
+                        "type": "object",
+                        "description": "The arguments to pass to the tool",
+                        "additionalProperties": True,
+                    },
+                },
+                "required": ["server_id", "tool_name", "parameters"],
+            },
+        })
+    
+    # Handle conversation and tool results
+    openai_messages = [{"role": "system", "content": system_content}]
+    
+    context_str = ""
+    for index, message in enumerate(request.messages):
+        if message.role == "human":
+            openai_messages.append({"role": "user", "content": message.content})
+        elif message.role == "ai":
+            if isinstance(message.content, str):
+                openai_messages.append({"role": "assistant", "content": message.content})
+        # Handle tool results in the conversation context
+        elif message.role == "tool" and index == len(request.messages) - 1:
+            context_str += "## MCP OUTPUT\n"
+            for result in message.data:
+                for item in result.items:
+                    context_str += f"{item.content}\n\n"
+            context_str += "## AI OUTPUT\n"
+            context_str += "Now provide your analysis based on the MCP output above.\n\n"
+    
+    if context_str:
+        openai_messages[-1]["content"] += "\n\n" + context_str
+    
+    async def execution_loop() -> AsyncGenerator[MessageChunkSSE | FunctionCallSSE, None]:
+        client = openai.AsyncOpenAI()
+        
+        # Check if we have tool results from previous execution
+        last_message = request.messages[-1] if request.messages else None
+        if last_message and last_message.role == "tool":
+            # Continue conversation with tool results
+            async for event in await client.chat.completions.create(
+                model="gpt-4o",
```

Comment:
As example using 4o feels old. I wonder if this code will work as is with 5.2-chat